src/server/models/db/migrations/20220913152735_missing_countries_lawyers/migration.sql

```diff
@@ -0,0 +1,114 @@
+-- Adds all remaining countries for lawyers service
```

Comment:
I'm not sure I follow. Are you referring to the comments?

Reply:
no, wasn't sure the best place to put it. although I probably overcomplicated it in my original comment. I think you can just do full outer joins for this, instead of having 3 different files that adds it for each service row by row. 

FULL OUTER JOIN will join two tables, where the condition doesn't match. For example, if there is a `List` with a `countryId`, it won't be returned.


Let's say we don't have any lists at all, but we do have all the countries. 
If you do a full outer join on the "Country" and "List", you will get 200 entries (or however many countries there are). It will include all of the columns from both entries (or you can pick which ones you want).

```

| id  | name                                         | countryId | serviceType | jsondata |
| --- | -------------------------------------------- | --------- | ----------- | -------- |
| 1   | Afghanistan                                  |           |             |          |
| 2   | Albania                                      |           |             |          

```

☝️ now you know which countries don't have a list. 

So, let's say we want to find all of the translator lists. 

```sql
SELECT * FROM "List" WHERE "type" = 'translator');
```

So combining this and the join, we can get all of the Countries that are missing a matching translator list. 

```sql
SELECT * FROM "List" WHERE "type" = 'translator') as translatorLists
               FULL OUTER JOIN "Country" ON "Country".id = "translatorLists"."countryId"
```

Now you want to "iterate" through this table, and insert a new list with the default jsonData.


```sql


INSERT INTO "List" ("countryId", "serviceType", "jsonData") -- add these columns to List. I've just simplified it here but you may need to add more cols if it complains
SELECT missingTranslatorCountries."id", 'translators', '{"the": "defaults"}' -- the values to add. missingTranslatorCountries."id" is added to the countryId column, 'translators' is added to serviceType etc 
FROM (SELECT "Country".id, --serviceType and jsonData is provided by the above line
      FROM (SELECT * FROM "List" WHERE "type" = 'translator') as translatorLists
               FULL OUTER JOIN "Country" ON "Country".id = "translatorLists"."countryId") as missingTranslatorCountries; 

```

Reply:
sure. if you let me know which part I can explain further. is it full outer join?

Reply:
sure. check slack 👍

Reply:
Took me a while to get my head round but I think I've managed to get it working. Thanks for the help 👍